### ISSUE TITLE
[API Proposal] Suggestions for a sensible Dataset API

### ISSUE THREAD

**Issue by PavlosMelissinos** [farizrahman4u/keras-contrib#153]
Hey there, so I've been silently working on a few ideas I've had about a sensible dataset format and I'm currently looking for some feedback before I open a PR here. Could someone check out [my version of the MS-COCO dataset](/PavlosMelissinos/enet-keras/blob/datasets_overhaul/src/data/datasets.py) and tell me what I would need to add/change in order to make it worthy of a merge?

For instance, it would make more sense if there were separate files for each dataset, like keras-contrib already is. I'm also in favor of a base, abstract dataset class with methods, like flow, and attributes like IDS/CLASS_NAMES that should be overriden, that needs some work, I've not yet decided how that would work.

I'm especially interested in your opinions regarding my use of a configuration file instead of feeding raw parameter values through code. I prefer my way visually and I think it's more readable but on the other hand it might be harder to debug.

Thanks a lot, I'm looking forward to your feedback!

**Comment by ahundt** [comment]
I think there are some valuable discussion points and pain points discussed in the tensorflow datasets api https://github.com/tensorflow/tensorflow/issues/7951, and interesting details on the kinds of transformations people have needed.

**Comment by ahundt** [comment]
the `_init_( config)` parameter would probably make the most sense simply being kwargs, which is how python handles all parameters to any function as a dictionary anyway.

**Comment by PavlosMelissinos** [comment]
Hey there, glad to see you so active! 👍

That tf thread seems handy, I could use the perspective.

> the _init_( config) parameter would probably make the most sense simply being kwargs, which is how python handles all parameters to any function as a dictionary anyway.

Well duh, indeed it makes perfect sense. I don't know why that didn't even cross my mind...

Thanks for the suggestions!

**Comment by ahundt** [comment]
@PavlosMelissinos It might be best to email the keras mailing list or post this in keras itself for more feedback

**Comment by ahundt** [comment]
@PavlosMelissinos is your version of the COCO dataset in a condition that can be merged here? My version isn't that great so it would be nice to pull it in!

**Comment by PavlosMelissinos** [comment]
@ahundt I've improved it a bit but it still needs a lot of cleaning up and I'm not actively working on it anymore. I want to make it more modular and there are a few things that I don't like. I've also pledged to contribute to a keras dataset project with @DEKHTIARJonathan which I've neglected for months (sorry Jonathan!). You can follow [that repo](https://github.com/DEKHTIARJonathan/keras-datasets) if you want and even offer your own implementation if you have one.

The actual implementation for the time being is still on my enet repo, along with some alternative network architectures but they all have to go.

**Comment by ahundt** [comment]
Ok I actually think I will need to use coco for some stuff and I already use keras-contrib for that so I'd like to just replace the current coco loading script with yours, rather than any of the models or anything. Plus, the coco data format supports other datasets too so it will be super convenient!

**Comment by PavlosMelissinos** [comment]
In that case feel free to take anything you want from my repo and modify it to your needs, I don't think I'll be doing any work on it for the next few days/weeks.

**Comment by PavlosMelissinos** [comment]
Oh, on a side note, I originally preferred using json files to define dataset parameters but not anymore. One of the first things I'd change now is turn it into code ("hard-coded" object constants) to remove a layer of (unnecessary) abstraction & complexity.

**Comment by ahundt** [comment]
Yeah I've found that "hard coding" is actually the way to go in python, as long as you separate the APIs so that you can change the settings programmatically too. Essentially python becomes your programming language and your config file language, then your commits naturally keep track of exactly what you ran. :-)
